src/core/server/base.js

```diff
@@ -112,6 +112,14 @@ function baseServer(routes, createStore, { appInstanceName = appName } = {}) {
       ) : {};
 
       const lang = getLangFromRouter(renderProps);
+
+      if (renderProps.params && renderProps.params.lang) {
+        const origLang = renderProps.params.lang;
+        if (lang !== origLang) {
+          return res.redirect(301, req.originalUrl.replace(encodeURI(origLang), lang));
```

Comment:
I'm not crazy about using replace. I feel like it might be better to explicitly replace the first part of the URL but I can't think of a great way of doing that. Splitting would be a bit gross but should work.